x-pack/plugins/apm/server/routes/agent_explorer/get_agents_items.ts

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { ProcessorEvent } from '@kbn/observability-plugin/common/processor_event';
+import {
+  kqlQuery,
+  rangeQuery,
+  termQuery,
+} from '@kbn/observability-plugin/server/utils/queries';
+import { map, uniq } from 'lodash';
+import {
+  AGENT_NAME,
+  AGENT_VERSION,
+  SERVICE_ENVIRONMENT,
+  SERVICE_LANGUAGE_NAME,
+  SERVICE_NAME,
+  SERVICE_NODE_NAME,
+} from '../../../common/elasticsearch_fieldnames';
+import { environmentQuery } from '../../../common/utils/environment_query';
+import { AgentName } from '../../../typings/es_schemas/ui/fields/agent';
+import { APMEventClient } from '../../lib/helpers/create_es_client/create_apm_event_client';
+import { RandomSampler } from '../../lib/helpers/get_random_sampler';
+import { MAX_NUMBER_OF_SERVICES } from '../services/get_services/get_services_items';
+
+interface AggregationParams {
+  environment: string;
+  serviceName?: string;
+  agentLanguage?: string;
+  kuery: string;
+  apmEventClient: APMEventClient;
+  start: number;
+  end: number;
+  randomSampler: RandomSampler;
+}
+
+export async function getAgentsItems({
+  environment,
+  agentLanguage,
+  serviceName,
+  kuery,
+  apmEventClient,
+  start,
+  end,
+  randomSampler,
+}: AggregationParams) {
+  const response = await apmEventClient.search('get_agent_details', {
+    apm: {
+      events: [ProcessorEvent.metric],
+    },
+    body: {
+      track_total_hits: false,
+      size: 0,
+      query: {
+        bool: {
+          filter: [
+            {
+              exists: {
+                field: AGENT_NAME,
+              },
+            },
+            {
+              exists: {
+                field: AGENT_VERSION,
+              },
+            },
+            ...rangeQuery(start, end),
+            ...environmentQuery(environment),
+            ...kqlQuery(kuery),
+            ...(serviceName ? termQuery(SERVICE_NAME, serviceName) : []),
+            ...(agentLanguage
+              ? termQuery(SERVICE_LANGUAGE_NAME, agentLanguage)
+              : []),
+          ],
+        },
+      },
+      aggs: {
+        sample: {
+          random_sampler: randomSampler,
+          aggs: {
+            services: {
+              terms: {
+                field: SERVICE_NAME,
+                size: MAX_NUMBER_OF_SERVICES,
+              },
+              aggs: {
+                instances: {
+                  cardinality: {
+                    field: SERVICE_NODE_NAME,
+                  },
+                },
+                serviceNodes: {
+                  terms: {
+                    field: SERVICE_NODE_NAME,
+                    size: MAX_NUMBER_OF_SERVICES,
```

Comment:
ah this is serviceNodes, not services. so `MAX_NUMBER_OF_SERVICES` isn't appropriate here. Additionally, you need to add `missing` here. We also need agent name/version/service language name for services without service.node.name (I missed the top metrics sub agg earlier when I mentioned we wouldn't need it).

Reply:
It looks like I cannot use `missing` when I'm using `sample`.
```
...
"error": {
    "root_cause": [
      {
        "type": "illegal_argument_exception",
        "reason": "[random_sampler] aggregation [sample] does not support sampling [missing] aggregation [noServiceNodeName]"
      },
...
```